src/binder/bind/bind_graph_pattern.cpp

```diff
@@ -27,12 +28,52 @@ Binder::bindGraphPattern(const std::vector<std::unique_ptr<PatternElement>>& gra
     return std::make_pair(std::move(queryGraphCollection), std::move(propertyCollection));
 }
 
+void static addUndirectedTableNames(const PatternElement& patternElement, int idx) {
+    if (patternElement.getNumPatternElementChains() == 0) {
+        return;
+    }
+    auto thisNode = idx == -1 ? patternElement.getFirstNodePattern() :
+                                patternElement.getPatternElementChain(idx)->getNodePattern();
+    // Left neighbor.
+    if (idx != -1) {
+        if (patternElement.getPatternElementChain(idx)->getRelPattern()->getDirection() ==
+            ArrowDirection::BOTH) {
+            auto leftNodeTableNames = idx == 0 ?
+                                          patternElement.getFirstNodePattern()->getTableNames() :
+                                          patternElement.getPatternElementChain(idx - 1)
+                                              ->getNodePattern()
+                                              ->getTableNames();
+            if (leftNodeTableNames.empty()) {
+                thisNode->clearTableNames();
+            } else {
+                thisNode->updateTableNames(leftNodeTableNames);
+            }
+        }
+    }
+    // Right neighbor.
+    if (patternElement.getNumPatternElementChains() <= idx + 1) {
+        return;
+    }
+    auto rightElementChain = patternElement.getPatternElementChain(idx + 1);
+    if (rightElementChain->getRelPattern()->getDirection() != ArrowDirection::BOTH) {
+        return;
+    }
+    auto rightNodeTableNames = rightElementChain->getNodePattern()->getTableNames();
+    if (rightNodeTableNames.empty()) {
+        thisNode->clearTableNames();
+    } else {
+        thisNode->updateTableNames(rightNodeTableNames);
+    }
+}
+
 // Grammar ensures pattern element is always connected and thus can be bound as a query graph.
 std::unique_ptr<QueryGraph> Binder::bindPatternElement(
     const PatternElement& patternElement, PropertyKeyValCollection& collection) {
     auto queryGraph = std::make_unique<QueryGraph>();
+    addUndirectedTableNames(patternElement, -1);
```

Comment:
Can we perform this rewriting/union logic by traversing `patternElement` twice?

If you look at the binding logic, at a high level it's looping through `patternElement` and keeping track of `leftNode` and `rightNode`
```
boundLeftNode = bind(first node pattern)
for (chain : patternElementChains) {
     boundRightNode = bind(chain.node)
     boundRel = bind(chain.rel)
     boundLeftNode = boundRightNode
}
```

We can do a similar rewriting/union logic and put it before binding. Something along the line of
```
leftNode = first node pattern
for (chain : patternElementChains) {
     rightNode = chain.node
     if (chain.rel.isUndirected) {
         tableNameUnion = unionTableName(leftNode, rightNode)
         leftNode.setTableNames(tableNameUnion);
         leftNode.setTableNames(tableNameUnion);
     }
}
```
To me this is more consistent and easier to understand. Also don't forget to add comments explaining why are we taking the union